src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -655,57 +661,45 @@ private Stream<JarEntry> jarStream() {
         synchronized (this) {
             ensureOpen();
             return StreamSupport.stream(new EntrySpliterator<>(0, res.zsrc.total,
-                pos -> (JarEntry)getZipEntry(null, pos)), false);
+                pos -> (JarEntry)getZipEntry(new EntryPos(getEntryName(pos), pos))), false);
         }
     }
 
     private String lastEntryName;
     private int lastEntryPos;
 
     /* Check ensureOpen() before invoking this method */
-    private ZipEntry getZipEntry(String name, int pos) {
+    private ZipEntry getZipEntry(EntryPos entryPos) {
+        String name = entryPos.name;
+        int pos  = entryPos.pos;
         byte[] cen = res.zsrc.cen;
-        int nlen = CENNAM(cen, pos);
-        int elen = CENEXT(cen, pos);
-        int clen = CENCOM(cen, pos);
+        ZipEntry e = this instanceof JarFile jarFile
+                ? Source.JUJA.entryFor(jarFile, name)
+                : new ZipEntry(name);
 
-        ZipCoder zc = res.zsrc.zipCoderForPos(pos);
-        if (name != null) {
-            // only need to check for mismatch of trailing slash
-            if (nlen > 0 &&
-                !name.isEmpty() &&
-                zc.hasTrailingSlash(cen, pos + CENHDR + nlen) &&
-                !name.endsWith("/"))
-            {
-                name += '/';
-            }
-        } else {
-            // invoked from iterator, use the entry name stored in cen
-            name = zc.toString(cen, pos + CENHDR, nlen);
-        }
-        ZipEntry e;
-        if (this instanceof JarFile) {
-            e = Source.JUJA.entryFor((JarFile)this, name);
-        } else {
-            e = new ZipEntry(name);
-        }
         e.flag = CENFLG(cen, pos);
+        e.method = CENHOW(cen, pos);
         e.xdostime = CENTIM(cen, pos);
         e.crc = CENCRC(cen, pos);
-        e.size = CENLEN(cen, pos);
         e.csize = CENSIZ(cen, pos);
-        e.method = CENHOW(cen, pos);
+        e.size = CENLEN(cen, pos);
         if (CENVEM_FA(cen, pos) == FILE_ATTRIBUTES_UNIX) {
             // read all bits in this field, including sym link attributes
             e.externalFileAttributes = CENATX_PERMS(cen, pos) & 0xFFFF;
         }
 
+        int elen = CENEXT(cen, pos);
+        int clen = CENCOM(cen, pos);
+
         if (elen != 0) {
+            int nlen = CENNAM(cen, pos);
             int start = pos + CENHDR + nlen;
             e.setExtra0(Arrays.copyOfRange(cen, start, start + elen), true, false);
         }
         if (clen != 0) {
+            int nlen = CENNAM(cen, pos);
```

Comment:
If this does not make a huge performance difference, I would keep this together with the init of elen and clen on line 692.

While there are not many  entry comments encountered, I do expect extra fields to somewhat common

Reply:
> If this does not make a huge performance difference, I would keep this together with the init of elen and clen on line 692.

I verified this does not cause a regression, so moved the `nlen` read up on the common path.

See also my comment to @cl4es re sequential memory access.